modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <iomanip>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace heph::telemetry {
+
+enum class Level : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
+auto operator<<(std::ostream& /*os*/, const Level& /*level*/) -> std::ostream&;
+
+template <typename T>
+concept NotQuotable = !std::convertible_to<T, std::string> && requires(std::ostream& os, T t) {
+  { os << t } -> std::same_as<std::ostream&>;
+};
+
+template <typename T>
+concept Quotable = std::convertible_to<T, std::string>;
+
+template <typename T>
+struct Field final {
+  std::string key;
+  T val;
+};
+
+/**
+ * @brief A class that allows easy composition of logs for structured logging.
+ *        Example(see also struclog.cpp):
+ *        namespace ht=heph::telemetry;
+ *        log(ht::LogEntry{Level::INFO, "adding"} | "id"_f(12345) | ht::Field{"tag", "test"}) ;
+ *
+ *        logs
+ *        'level=info file=struclog.h:123 time=2023-12-3T8:52:02+0 message="adding" id=12345 tag="test"'
+ *
+ *        Remark: We would like to use libfmt with quoted formatting as proposed in
+ *                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+ *
+ */
+class LogEntry {
+public:
+  using Fields = std::vector<Field<std::string>>;
+  using Clock = std::chrono::system_clock;
+
+public:
+  explicit LogEntry(Level level, std::string_view message,
+                    std::source_location location = std::source_location::current());
+
+  /*
+   * @brief General loginfo consumer, should be used like LogEntry("my message") | Field{"field", 1234}
+   *        Converted to string with stringstream.
+   *
+   * @tparam T any type that is consumeable by stringstream
+   * @param key identifier for the logging data
+   * @param val value of the logging data
+   * @return LogEntry&&
+   */
+  template <NotQuotable T>
+  auto operator|(const Field<T>& field) -> LogEntry&& {
+    std::stringstream ss;
+    ss << field.val;
+    fields.emplace_back(field.key, ss.str());
```

Comment:
normally I would agree however here I used it intentionally because:
A. it highlites that the Quotable and NonQuotable implementation are the same except the std::quoted
B. it would be the only placed used in this header and hence introduce a dep to fmt on the header (which is minor, since we use it everywhere, but still)...

Reply:
sounds good, we can keep it as it is